### PR TITLE
Misc updates for LDAP API authentication

### PR DESCRIPTION
**Pull request by mrichar1**
Although the docs say that creation of the salt-api user can be avoided by setting `SALT_API_USER=""` this doesn't currently work.

The code uses:

https://github.com/cdalvaro/docker-salt-master/blob/44c73c718b652b94b180c9214acdb9c8417e04b1/assets/runtime/env-defaults.sh#L11

However this doesn't distinguish between unset and null, so always sets `SALT_API_USER=salt_api`

This PR uses an alternative method of parameter expansion to test if `$SALT_API_USER` is undefined instead of null before setting a default value, resulting in:

```
> unset SALT_API_USER
> source runtime/env-defaults.sh
> echo $SALT_API_USER
salt_api
>
> SALT_API_USER=""
> source runtime/env-defaults.sh
> echo $SALT_API_USER

```

This then means that:

https://github.com/cdalvaro/docker-salt-master/blob/44c73c718b652b94b180c9214acdb9c8417e04b1/assets/runtime/functions.sh#L472

Will test false and skip user creation.

### CODE REVIEW

**Comment by cdalvaro** [comment]
Hi @mrichar1, thank you for opening this issue!

I see your point, but as long as `SALT_API_ENABLED` is set to anything other than `true`, the line you mentioned should never be executed:

https://github.com/cdalvaro/docker-salt-master/blob/44c73c718b652b94b180c9214acdb9c8417e04b1/assets/runtime/functions.sh#L470-L472

I’d like to better understand your issue. Are you simply expecting to see the `SALT_API_USER` environment variable empty when `SALT_API_ENABLED=false`, or have you encountered a specific failure?

**Comment by mrichar1** [comment]
Hi,

I'm trying to enable the API but then not create a local salt_api user as I'm planning on using LDAP for auth. So I'm setting:

```
SALT_API_ENABLED="True"
SALT_API_USER=""
```

However when I start the container it errors with:

```
[ERROR] - SALT_API_USER_PASS env variable must be set to create 'salt_api' user.
```

**Comment by cdalvaro** [comment]
OK, great!

So, I’d like to slightly change the purpose of the PR, if possible, and shift it more towards adding support for LDAP.

To achieve this, I’m going to ask you to:

- Get your modified image working with LDAP.
- Add documentation to the README with an example of how to configure `salt-api` with LDAP (focused only on the salt part; no need to provide details on setting up an LDAP server). I’ll handle the translation into Spanish.
- Add a test to verify that when the environment variable is set to empty, it doesn't default to a value inside the container.

Feel free to ask me for help with anything you need!

**Comment by mrichar1** [comment]
Yes I'm more than happy to work through the whole process and report back at the end. Will let you know how I get on!

**Comment by mrichar1** [comment]
I've added a couple more commits to get LDAP authentication working:

1. Add some build and run-time deps to the Docker image.
2. Document the steps to doing LDAP auth (with a simple example).

I'll come back again and have a look at writing a test for the empty `SALT_API_USER` soon.

Let me know if you have any comments or would like any changes made (especially if there's anything in the docs which isn't clear or needs more/less detail...)

File: Dockerfile [hunk 1]
```diff
@@ -33,11 +33,10 @@ RUN apt-get update \
   && DEBIAN_FRONTEND=noninteractive apt-get install --yes --quiet --no-install-recommends \
   sudo ca-certificates apt-transport-https wget locales openssh-client gpg gpg-agent \
   supervisor logrotate git gettext-base tzdata inotify-tools psmisc \
+  libldap-common gcc libsasl2-dev libldap2-dev \
```
Comment (cdalvaro) [inline Dockerfile:36]: I prefer to keep Dockerfile as short as possible and move these staff to [`assets/build/install.sh`](assets/build/install.sh)

There are two variable defined there:

- `REQUIRED_PACKAGES` where you can add `libldap-common`
- `BUILD_DEPENDENCIES` where you can add `gcc`, `libsasl2-dev`, `libldap2-dev`.

The second variable is clean at the end of the script.

https://github.com/cdalvaro/docker-salt-master/blob/44c73c718b652b94b180c9214acdb9c8417e04b1/assets/build/install.sh#L13-L17

File: Dockerfile [hunk 2]
```diff
@@ -48,8 +47,17 @@ RUN chmod -R +x ${SALT_RUNTIME_DIR}
...
+# FIXME: Downgrade to pip 22.3.1 for bug: https://github.com/saltstack/salt/issues/65025
+RUN salt-pip install "pip==22.3.1"
+RUN salt-pip install python-ldap
+
 # Cleaning tasks
 RUN rm -rf "${SALT_BUILD_DIR:?}"/*
+RUN DEBIAN_FRONTEND=noninteractive apt-get purge --yes --quiet gcc libsasl2-dev libldap2-dev \
+  && DEBIAN_FRONTEND=noninteractive apt-get --yes --quiet autoremove \
+  && DEBIAN_FRONTEND=noninteractive apt-get clean --yes \
+  && rm -rf /var/lib/apt/lists/*
```
Comment (cdalvaro) [inline Dockerfile:60]: This is not necessary, as it is already done here:

https://github.com/cdalvaro/docker-salt-master/blob/44c73c718b652b94b180c9214acdb9c8417e04b1/assets/build/install.sh#L116-L119

But you can add the `autoremove` command.

File: Dockerfile [hunk 3]
```diff
@@ -48,8 +47,17 @@ RUN chmod -R +x ${SALT_RUNTIME_DIR}
 
 COPY assets/sbin/* /usr/local/sbin/
 
+# Add python-ldap for ldap external_auth
+# FIXME: Downgrade to pip 22.3.1 for bug: https://github.com/saltstack/salt/issues/65025
+RUN salt-pip install "pip==22.3.1"
+RUN salt-pip install python-ldap
```
Comment (cdalvaro) [inline Dockerfile:53]: If there is an issue with the current `pip` version, I'm fine downgrading it.

But I'm not sure if I prefer to leave the `python-ldap` package optional for users who need it rather than installing it by default.

These users can install the package as a Python extra package: https://github.com/cdalvaro/docker-salt-master?tab=readme-ov-file#python-extra-packages

You should add to the LDAP documentation that this package has to be installed using this method in order to work.

On the other hand, if we install the package by default, y prefer to do it again in the [`assets/build/install.sh`](assets/build/install.sh) script.

https://github.com/cdalvaro/docker-salt-master/blob/44c73c718b652b94b180c9214acdb9c8417e04b1/assets/build/install.sh#L54

What do you think is better?

File: README.md [hunk 4]
```diff
@@ -351,6 +355,42 @@ external_auth:
...
+
+Finally (since `v3006`) you will need to enable one or more client interfaces:
+
+```yml
+netapi_enable_clients:
+  - local
+```
+
+Details of all client interfaces is available at the following
+link: [Netapi Client Interfaces](https://docs.saltproject.io/en/latest/topics/netapi/netapi-enable-clients.html)
```
Comment (cdalvaro) [inline README.md:392]: If we leave the `python-ldap` package optional, it would be great to add here a `compose.yml` example installing the package and setting the required env variables, if any.

File: Dockerfile (hunk 1 above)
Comment (mrichar1) [reply Dockerfile:36]: Oops - I missed that most of the work can happen in install.sh - I'll move the relevant bits over to it. I think that will mean that there wiull be no changes in `Dockerfile` in this PR at the end, so let me know if you're happy for me to rebase the last commit to it away, or would prefer an explicit 'revert' in the commit history.

Comment (cdalvaro) [reply Dockerfile:36]: I'm fine rebasing!

File: Dockerfile (hunk 3 above)
Comment (mrichar1) [reply Dockerfile:53]: I like the idea of making python-ldap optional - however installing this module requires quite large build deps (gcc etc) which would have to stay installed in the image for package install at run-time.

File: Dockerfile (hunk 2 above)
Comment (mrichar1) [reply Dockerfile:60]: Just checked and `install.sh` has `apt purge --auto-remove` and the man pages say: `this option acts like running the autoremove command` so probably don't need an explicit autoremove as well.

File: Dockerfile (hunk 3 above)
Comment (cdalvaro) [reply Dockerfile:53]: Yes, we need to keep the dependencies installed for the package to work. That’s why I wanted to discuss whether to make the installation of the python-ldap package optional or not.

If there had been no dependencies, we would have definitely made it optional.

Since the package shouldn’t take up too much space, I think we can install it by default.

**Comment by mrichar1** [comment]
That's the requested changes made and the PR rebased to remove the old commits. New changes are as follows:

1. `install.sh`: Add build deps and runtime deps for `python-ldap`, and install via `pip`
2. Test that container starts with `SALT_API_USER=""` (needed some changes to `tests/lib/common.h` to check the container exit code, and clean up exited containers).

The previous commits to allow empty `SALT_API_USER` and the documentation updates are unchanged.

File: tests/salt-api/test.sh [hunk 5]
```diff
@@ -41,6 +41,18 @@ EOF
 ok "salt-api config created"
 
 # Run test instance
+echo "==> Starting docker-salt-master (${PLATFORM}) with salt-api config and no api user ..."
+start_container_and_wait \
+  --publish 8000:8000 \
+  --env SALT_API_ENABLED=True \
+  --env SALT_API_USER="" \
+|| error "container started"
+ok "container started"
```
Comment (cdalvaro) [inline tests/salt-api/test.sh:50]: I would add one more check here:

```bash
check_equal "$(docker-exec salt-master bash -c 'env | grep SALT_API_USER= | cut -d= -f2')" "" "SALT_API_USER remains empty when explicitly defined that way"
```

**Comment by cdalvaro** [comment]
Would you be willing to create an integration test, `salt-api-ldap.sh`, using a test container, for example with `osixia/docker-openldap`?

It’s totally fine if it’s too much; I can add it later.

**Comment by cdalvaro** [comment]
Two more things:

- Update all references to revision `_3` pointing to revision `_4`. For example, inside the Dockerfile:

https://github.com/cdalvaro/docker-salt-master/blob/44c73c718b652b94b180c9214acdb9c8417e04b1/Dockerfile#L8

Other entries are: [README.md (EN)](docs/en-US/README.md), [README.md (ES)](docs/es-ES/README.md), and [bug_report.md](.github/ISSUE_TEMPLATE/bug_report.md)

- Update [CHANGELOG.md](CHANGELOG.mg) adding and entry for this new version. Something like:

```md
**3007.1_4**

- Add support for LDAP ([#267](https://github.com/cdalvaro/docker-salt-master/pull/267))
```

File: tests/salt-api/test.sh (hunk 5 above)
Comment (mrichar1) [reply tests/salt-api/test.sh:50]: Good idea - added in rebased commit

**Comment by mrichar1** [comment]
I'm happy to have a look at an integration test - though it might be better putting in a fresh PR as it might take a bit longer and need more discussion (e.g. applying test functions like `start_container_and_wait` and `cleanup` to 3rd party docker containers etc).

**Comment by cdalvaro** [comment]
> I'm happy to have a look at an integration test - though it might be better putting in a fresh PR as it might take a bit longer and need more discussion (e.g. applying test functions like `start_container_and_wait` and `cleanup` to 3rd party docker containers etc).

OK! It seems right!

**Comment by cdalvaro** [comment]
> Two more things:
> 
> * Update all references to revision `_3` pointing to revision `_4`. For example, inside the Dockerfile:
> 
> https://github.com/cdalvaro/docker-salt-master/blob/44c73c718b652b94b180c9214acdb9c8417e04b1/Dockerfile#L8
> 
> Other entries are: `docs/*/README.md` and `bug_report.md`.
> 
> * Update [CHANGELOG.md](CHANGELOG.mg) adding and entry for this new version. Something like:
> 
> ```
> **3007.1_4**
> 
> - Add support for LDAP ([#267](https://github.com/cdalvaro/docker-salt-master/pull/267))
> ```

Hi @mrichar1, just a reminder. As soon as you make these changes I'll approve the PR.

**Comment by mrichar1** [comment]
That's the version info updated everywhere :+1:

**Comment by cdalvaro** [comment]
Thank you very much, @mrichar1! Great contribution! 🚀
